mutiny-core/src/nostr/nwc.rs

```diff
@@ -1032,26 +1135,319 @@ impl NostrWalletConnect {
+                    None => event.id.to_string(),
                 };
+                let preimage = PaymentPreimage(Sha256::hash(input.as_bytes()).into_32());
+                let payment_hash = PaymentHash(Sha256::hash(&preimage.0).into_32());
+
+                // skip if keysend payment has already been done or in-flight
+                if let Some(payment) = node.lookup_payment(&payment_hash.0).await {
+                    if payment.status == HTLCStatus::Succeeded
+                        || payment.status == HTLCStatus::InFlight
+                    {
+                        log_warn!(nostr_manager.logger, "keysend payment already made");
+                        return Ok(None);
+                    }
+                }
 
-                let response = EventBuilder::new(Kind::WalletConnectResponse, encrypted, tags)
-                    .to_event(&self.server_key)?;
+                let sats = params.amount / 1000;
 
-                Ok(Some(response))
+                // now only try keysend payment if within budget
+                // if it falls out of budget return, send error
+                let content = match self
+                    .check_payment_within_budget(node, &mut budget, sats)
+                    .await
+                {
+                    Err(err) => {
+                        log_warn!(nostr_manager.logger, "Attempted to exceed budget: {err}");
+                        Response {
+                            result_type: method,
+                            error: Some(NIP47Error {
+                                code: ErrorCode::QuotaExceeded,
+                                message: err,
+                            }),
+                            result: None,
+                        }
+                    }
+                    Ok(_) => {
+                        let mut payment = TrackedPayment {
+                            time: utils::now().as_secs(),
+                            amt: sats,
+                            hash: payment_hash.to_string(),
+                        };
+                        // add payment to budget
+                        budget.add_payment(&mut payment);
+                        self.profile.spending_conditions =
+                            SpendingConditions::Budget(budget.clone());
+                        // persist budget before payment to protect against it not saving after
+                        nostr_manager.save_nwc_profile(self.clone())?;
+
+                        let to_node_pubkey = PublicKey::from_str(params.pubkey.as_str())?;
+
+                        let label = self
+                            .profile
+                            .label
+                            .clone()
+                            .unwrap_or(self.profile.name.clone());
+                        // attempt keysend payment now
+                        match node
+                            .keysend(to_node_pubkey, sats, None, vec![label], Some(preimage.0))
```

Comment:
you're never doing anything with the TLV records will need to add that to the keysend function too

Reply:
@TonyGiorgio passing that optional preimage here

Reply:
@benthecarman did some changes to include the TLV records. I changed the previous message param and passing new `Vec<CustomTLV>`. This would require changes in the frontend tho. I saw the frontend was not passing a message anyways so it would be changing from undefined to an empty array. Let me know what you think